store/src/java/com/zimbra/cs/service/ExternalUserProvServlet.java

```diff
@@ -277,19 +278,70 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws S
 //        String folderId = (String) tokenMap.get("fid");
         String extUserEmail = (String) tokenMap.get("email");
 
-        provisionVirtualAccountAndRedirect(req, resp, displayName, password, ownerId, extUserEmail);
+        try {
+            provisionVirtualAccountAndRedirect(req, resp, displayName, password, ownerId, extUserEmail);
+        } catch (ServiceException e) {
+            Map<String, String> errorAttrs = new HashMap<String, String>();
+            errorAttrs.put("errorCode", e.getCode());
+            errorAttrs.put("errorMessage", e.getMessage());
+            redirectError(req, resp, errorAttrs);
+        } catch (Exception e) {
+            Map<String, String> errorAttrs = new HashMap<String, String>();
+            errorAttrs.put("errorCode", ServiceException.FAILURE);
+            errorAttrs.put("errorMessage", e.getMessage());
+            redirectError(req, resp, errorAttrs);
+        }
+    }
+
+    private void redirectError(HttpServletRequest req, HttpServletResponse resp, Map<String, String> errorAttrs)
+        throws ServletException, IOException {
+        if (WebClientServiceUtil.isServerInSplitMode()) {
+            Map<String, String> reqHeaders = new HashMap<String, String>();
+            for (Map.Entry<String, String> entry : errorAttrs.entrySet()) {
+                reqHeaders.put(entry.getKey(), entry.getValue());
+            }
+            sendHtmlResponse(resp, reqHeaders);
+        } else {
+            for (Map.Entry<String, String> entry : errorAttrs.entrySet()) {
+                req.setAttribute(entry.getKey(), entry.getValue());
+            }
+            forward(req, resp);
+        }
+    }
+
+    private void forward(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
```

Comment:
similar code is there at several places in this class, better to make the jsp page as argument of the method and use this method at all these places to remove duplicate code. same for sendHtmlResposne method.